components/feature/session/src/main/java/mozilla/components/feature/session/DefaultSessionStorage.kt

```diff
@@ -84,43 +58,35 @@ class DefaultSessionStorage(private val context: Context) : SessionStorage {
                     val jsonSession = jsonRoot.getJSONObject(it)
                     val session = deserializeSession(jsonSession.getJSONObject(SESSION_KEY))
                     val engineSession = deserializeEngineSession(engine, jsonSession.getJSONObject(ENGINE_SESSION_KEY))
-                    sessions.put(it, SessionProxy(session, engineSession))
+                    sessions.put(session, engineSession)
                 }
-                selectedSession = sessions[selectedSessionId]?.session
+                Pair(sessions, selectedSessionId)
             }
-            sessions.values.toList()
-        } catch (_: FileNotFoundException) {
-            emptyList()
+        } catch (_: Throwable) {
+            Pair(emptyMap(), "")
         }
     }
 
     @Synchronized
-    override fun persist(selectedSession: Session?): Boolean {
+    @Suppress("TooGenericExceptionCaught")
+    override fun persist(sessions: Map<Session, EngineSession>, selectedSession: String): Boolean {
         return try {
             val json = JSONObject()
             json.put(VERSION_KEY, VERSION)
+            json.put(SELECTED_SESSION_KEY, selectedSession)
 
-            var selectedSessionId = ""
-            for ((key, value) in sessions) {
-                if (value.session == selectedSession) {
-                    selectedSessionId = key
-                }
-            }
-            json.put(SELECTED_SESSION_KEY, selectedSessionId)
-
-            sessions.forEach({ (key, value) ->
+            sessions.forEach({ (session, engineSession) ->
                 val sessionJson = JSONObject()
-                sessionJson.put(SESSION_KEY, serializeSession(value.session))
-                sessionJson.put(ENGINE_SESSION_KEY, serializeEngineSession(value.engineSession))
-                json.put(key, sessionJson)
+                sessionJson.put(SESSION_KEY, serializeSession(session))
+                sessionJson.put(ENGINE_SESSION_KEY, serializeEngineSession(engineSession))
+                json.put(session.id, sessionJson)
             })
 
             context.openFileOutput(FILE_NAME, Context.MODE_PRIVATE).use {
-                println(json.toString())
                 it.write(json.toString().toByteArray())
             }
             true
-        } catch (_: FileNotFoundException) {
+        } catch (_: Throwable) {
```

Comment:
I wonder why we catch Throwable now? I'm always afraid that we accidentally add things like NullPointerExceptions and then not notice them.

Reply:
@pocmo I wanted a catch all to prevent any exception from bubbling up and preventing startup i.e. if the file is not found, or the JSON malformed. Unfortunately, Kotlin doesn't have multi catch yet, but I will refactor to catch FileNotFound and JSON exceptions directly.

Reply:
Oh, wow. That's the first language feature where Java is superior :D